File: plugins/plugin-bootstrap/python/elizaos_plugin_bootstrap/actions/unmute_room.py
"""
UNMUTE_ROOM Action - Unmute a room to resume receiving notifications.

This action allows the agent to unmute a previously muted room
and resume responding to messages.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content, MemoryType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class UnmuteRoomAction:
    """
    Action for unmuting a room.

    This action is used when:
    - The agent should resume responding to room messages
    - Notifications from a room should be enabled
    - The agent wants to re-engage with a room
    """

    name: str = "UNMUTE_ROOM"
    similes: list[str] = field(
        default_factory=lambda: [
            "UNSILENCE_ROOM",
            "ENABLE_NOTIFICATIONS",
            "RESUME_RESPONDING",
            "START_LISTENING",
        ]
    )
    description: str = (
        "Unmute a room to resume responding and receiving notifications. "
        "Use this when you want to start interacting with a muted room again."
    )

    async def validate(self, runtime: IAgentRuntime, message: Memory) -> bool:
        """Validate that room is currently muted."""
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None:
            return False

        # Check if currently muted
        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                muted_rooms = world.metadata.get("mutedRooms", [])
                # Only valid if currently muted
                if str(room_id) in muted_rooms:
                    return True

        return False

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        """Handle unmuting a room."""
        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room specified to unmute",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "UNMUTE_ROOM"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None:
            return ActionResult(
                text="Room not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "UNMUTE_ROOM"},
                success=False,
            )

        room_name = str(room.name) if room.name else "Unknown Room"

        try:
            # Get world and update muted rooms
            world_id = room.world_id
            if world_id:
                world = await runtime.get_world(world_id)
                if world and world.metadata:
                    muted_rooms = list(world.metadata.get("mutedRooms", []))
                    room_id_str = str(room_id)

                    if room_id_str in muted_rooms:
                        muted_rooms.remove(room_id_str)
                        world.metadata["mutedRooms"] = muted_rooms
                        await runtime.update_world(world)

            # Create memory of the action
            await runtime.create_memory(
                content=Content(
                    text=f"Unmuted room: {room_name}",
                    actions=["UNMUTE_ROOM"],
                ),
                room_id=room_id,
                entity_id=runtime.agent_id,
                memory_type=MemoryType.ACTION,
                metadata={"type": "UNMUTE_ROOM", "roomName": room_name},
            )

            response_content = Content(
                text=f"I have unmuted {room_name}. I will now respond to messages there.",
                actions=["UNMUTE_ROOM"],
            )

            if callback:
                await callback(response_content)

            return ActionResult(
                text=f"Unmuted room: {room_name}",
                values={
                    "success": True,
                    "unmuted": True,
                    "roomId": str(room_id),
                    "roomName": room_name,
                },
                data={
                    "actionName": "UNMUTE_ROOM",
                    "roomId": str(room_id),
                    "roomName": room_name,
                },
                success=True,
            )

        except Exception as error:
            runtime.logger.error(
                {
                    "src": "plugin:bootstrap:action:unmuteRoom",
                    "agentId": runtime.agent_id,
                    "roomId": str(room_id),
                    "error": str(error),
                },
                "Error unmuting room",
            )
            return ActionResult(
                text="Error unmuting room",
                values={"success": False, "error": str(error)},
                data={"actionName": "UNMUTE_ROOM", "error": str(error)},
                success=False,
                error=error,
            )

    @property
    def examples(self) -> list[list[ActionExample]]:
        """Example interactions demonstrating the UNMUTE_ROOM action."""
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="You can start responding in this channel again."),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll unmute this room and start responding again.",
                        actions=["UNMUTE_ROOM"],
                    ),
                ),
            ],
        ]


# Create the action instance
unmute_room_action = Action(
    name=UnmuteRoomAction.name,
    similes=UnmuteRoomAction().similes,
    description=UnmuteRoomAction.description,
    validate=UnmuteRoomAction().validate,
    handler=UnmuteRoomAction().handler,
    examples=UnmuteRoomAction().examples,
)

File: packages/typescript/src/advanced-capabilities/actions/unmuteRoom.ts
import { requireActionSpec } from "../../generated/spec-helpers.ts";
import { logger } from "../../logger.ts";
import { shouldUnmuteRoomTemplate } from "../../prompts.ts";
import type {
  Action,
  ActionExample,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "../../types/index.ts";
import { ModelType } from "../../types/index.ts";
import { composePromptFromState } from "../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("UNMUTE_ROOM");

export const unmuteRoomAction: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  examples: (spec.examples ?? []) as ActionExample[][],
  validate: async (runtime: IAgentRuntime, message: Memory) => {
    const roomId = message.roomId;
    const roomState = await runtime.getParticipantUserState(
      roomId,
      runtime.agentId,
    );
    return roomState === "MUTED";
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    _callback?: HandlerCallback,
    _responses?: Memory[],
  ): Promise<ActionResult> => {
    async function _shouldUnmute(state: State): Promise<boolean> {
      const shouldUnmutePrompt = composePromptFromState({
        state,
        template: shouldUnmuteRoomTemplate,
      });

      const response = await runtime.useModel(ModelType.TEXT_SMALL, {
        prompt: shouldUnmutePrompt,
        stopSequences: [],
      });

      const cleanedResponse = response.trim().toLowerCase();

      // Handle various affirmative responses
      if (
        cleanedResponse === "true" ||
        cleanedResponse === "yes" ||
        cleanedResponse === "y" ||
        cleanedResponse.includes("true") ||
        cleanedResponse.includes("yes")
      ) {
        await runtime.createMemory(
          {
            entityId: message.entityId,
            agentId: message.agentId,
            roomId: message.roomId,
            content: {
              source: message.content.source,
              thought:
                "I will now unmute this room and start considering it for responses again",
              actions: ["UNMUTE_ROOM_STARTED"],
            },
          },
          "messages",
        );
        return true;
      }

      // Handle various negative responses
      if (
        cleanedResponse === "false" ||
        cleanedResponse === "no" ||
        cleanedResponse === "n" ||
        cleanedResponse.includes("false") ||
        cleanedResponse.includes("no")
      ) {
        await runtime.createMemory(
          {
            entityId: message.entityId,
            agentId: message.agentId,
            roomId: message.roomId,
            content: {
              source: message.content.source,
              thought: "I tried to unmute a room but I decided not to",
              actions: ["UNMUTE_ROOM_FAILED"],
            },
          },
          "messages",
        );
        return false;
      }

      // Default to false if response is unclear
      logger.warn(
        {
          src: "plugin:bootstrap:action:unmute_room",
          agentId: runtime.agentId,
          response,
        },
        "Unclear boolean response, defaulting to false",
      );
      return false;
    }

    if (!state) {
      return {
        text: "State is required for unmute room action",
        values: {
          success: false,
          error: "STATE_REQUIRED",
        },
        data: {
          actionName: "UNMUTE_ROOM",
          error: "State is required",
        },
        success: false,
        error: new Error("State is required for unmute room action"),
      };
    }

    const shouldUnmute = await _shouldUnmute(state);

    if (shouldUnmute) {
      try {
        await runtime.setParticipantUserState(
          message.roomId,
          runtime.agentId,
          null,
        );

        const room = await runtime.getRoom(message.roomId);

        if (!room) {
          logger.warn(
            {
              src: "plugin:bootstrap:action:unmute_room",
              agentId: runtime.agentId,
              roomId: message.roomId,
            },
            "Room not found",
          );
          return {
            text: `Room not found: ${message.roomId}`,
            values: {
              success: false,
              error: "ROOM_NOT_FOUND",
              roomId: message.roomId,
            },
            data: {
              actionName: "UNMUTE_ROOM",
              error: "Room not found",
              roomId: message.roomId,
            },
            success: false,
          };
        }

        // Ensure room has a name for consistent return values
        const roomName = room.name ?? `Room-${message.roomId.substring(0, 8)}`;

        await runtime.createMemory(
          {
            entityId: message.entityId,
            agentId: message.agentId,
            roomId: message.roomId,
            content: {
              thought: `I unmuted the room ${roomName}`,
              actions: ["UNMUTE_ROOM_START"],
            },
          },
          "messages",
        );

        return {
          text: `Room unmuted: ${roomName}`,
          values: {
            success: true,
            roomUnmuted: true,
            roomId: message.roomId,
            roomName: roomName,
            newState: "NONE",
          },
          data: {
            actionName: "UNMUTE_ROOM",
            roomId: message.roomId,
            roomName: roomName,
            unmuted: true,
          },
          success: true,
        };
      } catch (error) {
        logger.error(
          {
            src: "plugin:bootstrap:action:unmute_room",
            agentId: runtime.agentId,
            error: error instanceof Error ? error.message : String(error),
          },
          "Error unmuting room",
        );
        return {
          text: "Failed to unmute room",
          values: {
            success: false,
            error: "UNMUTE_FAILED",
          },
          data: {
            actionName: "UNMUTE_ROOM",
            error: error instanceof Error ? error.message : String(error),
            roomId: message.roomId,
          },
          success: false,
          error: error instanceof Error ? error : new Error(String(error)),
        };
      }
    } else {
      return {
        text: "Decided not to unmute room",
        values: {
          success: true,
          roomUnmuted: false,
          roomId: message.roomId,
          reason: "CRITERIA_NOT_MET",
        },
        data: {
          actionName: "UNMUTE_ROOM",
          roomId: message.roomId,
          unmuted: false,
          reason: "Decision criteria not met",
        },
        success: true,
      };
    }
  },
};
